addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/handler/MySensorsThingHandler.java

```diff
@@ -0,0 +1,396 @@
+                        newMsg.setMsg(adapter.fromCommand(command));
+
+                        myGateway.sendMessage(newMsg);
+
+                    } else {
+                        logger.warn("Variable not found, cannot handle command for thing {} of type {}", thing.getUID(),
+                                channelUID.getId());
+                    }
+                } else {
+                    logger.error("Could not get type of variable for channel: {}, command: {} of thing {}",
+                            thing.getUID(), command, thing.getUID());
+                }
+
+            } else {
+                logger.error("Type adapter not found for {}", channelUID.getId());
+            }
+
+        }
+    }
+
+    @Override
+    public void messageReceived(MySensorsMessage message) throws Exception {
+        handleIncomingMessageEvent(message);
+
+    }
+
+    @Override
+    public void sensorUpdateEvent(MySensorsNode node, MySensorsChild child, MySensorsVariable var,
+            MySensorsNodeUpdateEventType eventType) {
+        switch (eventType) {
+            case UPDATE:
+            case REVERT:
+                if ((node.getNodeId() == nodeId) && (child.getChildId() == childId)) {
+                    handleChildUpdateEvent(var);
+                    updateLastUpdate(node, eventType == MySensorsNodeUpdateEventType.REVERT);
+                }
+                break;
+            case BATTERY:
+                if (node.getNodeId() == nodeId) {
+                    handleBatteryUpdateEvent(node);
+                    updateLastUpdate(node, eventType == MySensorsNodeUpdateEventType.REVERT);
+                }
+                break;
+        }
+
+    }
+
+    @Override
+    public void connectionStatusUpdate(MySensorsAbstractConnection connection, boolean connected) throws Exception {
+        if (!connected) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+        } else {
+            updateStatus(ThingStatus.ONLINE);
+        }
+    }
+
+    @Override
+    public void nodeReachStatusChanged(MySensorsNode node, boolean reach) {
+        if (node.getNodeId() == nodeId) {
+            if (!reach) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            } else {
+                updateStatus(ThingStatus.ONLINE);
+            }
+        }
+
+    }
+
+    /**
+     * For every thing there is a lastUpdate channel in which the date/time is stored
+     * a message was received from this thing.
+     */
+    private void updateLastUpdate(MySensorsNode node, boolean isRevert) {
+        // Don't always fire last update channel, do it only after a minute by
+        if (lastUpdate == null || (System.currentTimeMillis() > (lastUpdate.getCalendar().getTimeInMillis() + 60000))
+                || revertState) {
+            DateTimeType dt = new DateTimeType(
+                    new SimpleDateFormat(DateTimeType.DATE_PATTERN).format(node.getLastUpdate()));

```

Comment:
SimpleDateFormat is quite slow (and not thread-safe) I don't know how many nodes (and updates) a common network has but it good to keep that in mind